homeassistant/components/unifiprotect/data.py

```diff
@@ -188,3 +214,40 @@ def async_signal_device_id_update(self, device_id: str) -> None:
         _LOGGER.debug("Updating device: %s", device_id)
         for update_callback in self._subscriptions[device_id]:
             update_callback()
+
+    @callback
+    def _async_get_device_from_ref(
+        self, ref: ProtectDeviceRef
+    ) -> ProtectAdoptableDeviceModel | NVR | None:
+
+        if ref.model == ModelType.NVR:
+            if ref.device_id == self.api.bootstrap.nvr.id:
+                return self.api.bootstrap.nvr
+            return None  # pragma: no cover
+
+        device = getattr(self.api.bootstrap, f"{ref.model}s").get(ref.device_id)
+        return cast(Optional[ProtectAdoptableDeviceModel], device)
+
+    @callback
+    def async_get_ufp_device_from_device(
+        self, device_id: str
+    ) -> ProtectAdoptableDeviceModel | NVR | None:
+        """Get UniFi Protect device from HA device ID."""
+
+        ref = self._device_to_device_map.get(device_id)
+        if ref is None:
+            return None  # pragma: no cover
```

Comment:
```suggestion
        ref = self._device_to_device_map[device_id]
```

If this is unreachable, we should not use a get

Reply:
It is not that it is unreachable, but rather untestable given the HA core testing "standards". If you call this with a non-existent device ID without a get, it _will_ raise an `IndexError`. There is no code that _currently_ does that, but leaving in a code path could result in a user facing error is not good.

Reply:
I don't understand how its untestable if a user can cause it?

Reply:
* An AppDaemon or other python code that can directly interface with `hass.data`
* Service call validation changes and an invalid device ID is passed in from a service call
* New code is added in the future that constructs the entity or device ID in some other method
* Race condition from calling a service and deleting entity?